[Demangler][Tests] Add an explicit test for getObjCClassByMangledName.

al45tair opened the pull request:
Added a special test for `getObjCClassByMangledName`; this needs testing separately as it uses the `DecodedMetadataBuilder`, which doesn't get exercised by the normal demangling tests.

Added all the test cases from rdar://63485806, rdar://63488139, rdar://63496478, rdar://63410196 and rdar://68449341.  The test cases from rdar://63485806 are disabled for now because the problem there is the error handling mechanism (or
lack thereof), rather than us not handling errors.

Fixes the remaining cases from

rdar://63488139
rdar://63496478

al45tair committed:
[Demangler][Tests] Add an explicit test for getObjCClassByMangledName.

Added a special test for getObjCClassByMangledName; this needs testing
separately as it uses the DecodedMetadataBuilder, which doesn't get exercised
by the normal demangling tests.

Added all the test cases from rdar://63485806, rdar://63488139, rdar://63496478,
rdar://63410196 and rdar://68449341.  The test cases from rdar://63485806 are
disabled for now because the problem there is the error handling mechanism (or
lack thereof), rather than us not handling errors.

Fixes the remaining cases from

rdar://63488139
rdar://63496478

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented on lib/Demangling/Remangler.cpp:
```diff
@@ -441,9 +441,11 @@ std::pair<int, Node *> Remangler::mangleConstrainedType(Node *node,
   if (node->getKind() != Node::Kind::DependentGenericParamType) {
     mangle(node, depth + 1);
+    if (!Chain.size())
+      return {-1, nullptr};
```
This bit slightly concerns me. It makes sense, and it passes all the tests, but I'm not 100% confident of it. I *think* it's right because if there are no dependent member types and no dependent generic parameter types, it can't be a constrained type and should be rejected.

Without the `if (!Chain.size())`, we break `Remangler::mangleDependentMemberType`, where it's clearly legal for the returned `Node *` to be `nullptr`. And without the `return` statement here at all, we can be tripped up in `Remangler::mangleDependentGenericConformanceRequirement()`, `Remangler::mangleDependentGenericSameTypeRequirement()` or `Remangler::mangleDependentGenericLayoutRequirement()`; the test vectors that accompany these changes trigger the assertion failure in the second of those.

DougGregor replied:
There are no dependent member types that don't eventually terminate at a generic parameter, so this `if` makes sense for detecting a mangle tree that describes a type that is structurally impossible in a well-formed program.

al45tair replied:
Thanks for confirming that. 🙂

DougGregor commented on stdlib/public/runtime/Metadata.cpp:
```diff
@@ -3484,6 +3484,10 @@ ExistentialMetatypeValueWitnessTableCacheEntry(unsigned numWitnessTables) {
 SWIFT_RUNTIME_EXPORT
 const ExistentialMetatypeMetadata *
 swift::swift_getExistentialMetatypeMetadata(const Metadata *instanceMetadata) {
+  if (instanceMetadata->getKind() != MetadataKind::Existential
+      && instanceMetadata->getKind() != MetadataKind::ExistentialMetatype)
+    return nullptr;
```
All of the other (non-metatype-decoding) callers to `swift_getExistentialMetatypeMetadata` already ensure that they provide correct instance metadata, so this check is redundant and the result can be assumed to never be NULL. I think this check should be lifted up into the caller(s) in the demangle-to-metadata path.

al45tair replied:
Done.

DougGregor commented on stdlib/public/runtime/MetadataLookup.cpp:
```diff
@@ -1446,7 +1446,13 @@ class DecodedMetadataBuilder {
       BuiltType instance,
       llvm::Optional<Demangle::ImplMetatypeRepresentation> repr = None) const {
-    return swift_getExistentialMetatypeMetadata(instance);
+    BuiltType result = swift_getExistentialMetatypeMetadata(instance);
```
(Here's where I think the `Existential`/`ExistentialMetatype` check should go)

DougGregor reviewed (approved):
Some minor comments, but LGTM!

al45tair committed:
[Demangler][Tests] Address Doug's comments.

Moved the test for the metadata kind to MetadataLookup.cpp.

Added an assertion (for debug builds) to Metadata.cpp to catch the case where
something manages to bypass that test.
